### PR TITLE
test-to-harness: initial set up

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Ref: https://github.com/google/oss-fuzz-gen/issues/494

Some more comments on this PR in https://github.com/google/oss-fuzz-gen/pull/511#issuecomment-2254155947

### CODE REVIEW

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-test-to-harenss-1 -m vertex_ai_gemini-1-5 -b comparison from-test-small

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-test-to-harenss-2 -m vertex_ai_gemini-1-5 -b from-test-small

**Comment by DavidKorczynski** [comment]
Experiment from test-to-harness conversion: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-07-27-511-dk-test-to-harenss-2-from-test-small/index.html

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-1231231432-m vertex_ai_gemini-1-5 -b comparison

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-comparison-jj1 -m vertex_ai_gemini-1-5 -b comparison

**Comment by DavidKorczynski** [comment]
regular comparison benchmark: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-07-27-511-dk-comparison-jj1-comparison//index.html

**Comment by DavidKorczynski** [comment]
This is ready for review. The experiment https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-07-27-511-dk-test-to-harenss-2-from-test-small/index.html shows improvements in several projects where we previously had no gains: https://llm-exp.oss-fuzz.com/Result-reports/scheduled/2024-07-06-weekly-all/index.html 

I think this is a promising direction, not least because we're seeing improvements and there are many technical improvements we can do since we now, more or less, only do a copy-paste of the test into the prompt. I think the PR is in a state though where we can do incremental improvements on this.

Particularly I think there are improvements needed in (1) architecture around benchmarks; (2) more context around tests; (3) more experiments around tests, e.g. we copy whole files in now where we could probably refine this (e.g. where there are multiple tests in a file we can extract the tests).

File: data_prep/introspector.py [hunk 1]
```diff
@@ -650,6 +662,38 @@ def populate_benchmarks_using_introspector(project: str, language: str,
...
+                                   function_name='test-file',
+                                   return_type='test',
+                                   params=[],
+                                   exceptions=[],
+                                   is_jvm_static=False,
+                                   target_path=harness,
+                                   preferred_target_name='',
+                                   is_test_benchmark=True,
+                                   test_file_path=test_file))
+      return potential_benchmarks[:limit]
```
Comment (DonggeLiu) [inline data_prep/introspector.py:695]: Shall we do this in its own function and call the function under `if __name__ == '__main__':`, given this part is very different and independent from the rest of this function?
E.g.,
```python
def _populate_benchmarks_using_test_migration(...):
  ...

if __name__ == '__main__':
  ...
  if 'test-migration' in target_oracle:
    benchmarks = populate_benchmarks_using_test_migration(...)
```

File: experiment/benchmark.py [hunk 2]
```diff
@@ -83,32 +86,60 @@ def from_yaml(cls, benchmark_path: str) -> List:
...
+                function.get('params'),
+                function.get('exceptions', []),
+                function.get('is_jvm_static', False),
+                data['target_path'],
+                data.get('target_name'),
+                use_project_examples=use_project_examples,
+                cppify_headers=cppify_headers,
+                commit=commit,
+                use_context=use_context,
+                function_dict=function))
```
Comment (DonggeLiu) [inline experiment/benchmark.py:142]: OK for now, but would you please merge the same code in if/else block later to reduce repetition later?
Thanks

File: llm_toolkit/prompt_builder.py [hunk 3]
```diff
@@ -1125,3 +1125,112 @@ def post_process_generated_code(self, generated_code: str) -> str:
...
+
+  def build_triager_prompt(self, benchmark: Benchmark, driver_code: str,
+                           crash_info: str, crash_func: dict) -> prompts.Prompt:
+    """Builds a triager prompt."""
+    return self._prompt
+
+  def post_process_generated_code(self, generated_code: str) -> str:
+    """Adds specific C headers we always want in the harnesses."""
+    for header in C_PROMPT_HEADERS_TO_ALWAYS_INCLUDES:
+      generated_code = f'#include <{header}>\n' + generated_code
```
Comment (DonggeLiu) [inline llm_toolkit/prompt_builder.py:1231]: nit: `generated_code = f'#include <{header}>\n{generated_code}'`

File: run_all_experiments.py [hunk 4]
```diff
@@ -156,7 +155,7 @@ def run_experiments(benchmark: benchmarklib.Benchmark,
         prompt_builder_to_use=args.prompt_builder)
     return Result(benchmark, result)
   except Exception as e:
-    logger.error('Exception while running experiment: %s', str(e))
+    logger.info('Exception while running experiment: %s', str(e))
```
Comment (DonggeLiu) [inline run_all_experiments.py:158]: Why do we change this?

File: run_one_experiment.py [hunk 5]
```diff
@@ -243,17 +243,25 @@ def generate_targets_for_analysis(model: models.LLM,
...
+  # If this is a test benchmark then we will use a test prompt builder.
+  if benchmark.is_test_benchmark:
+    logger.info('Generating a target for test case: %s',
+                benchmark.test_file_path)
+    builder = prompt_builder.TestToHarnessConverter(model, benchmark,
+                                                    template_dir)
   else:
-    if prompt_builder_to_use == 'CSpecific':
-      builder = prompt_builder.CSpecificBuilder(model, benchmark, template_dir)
+    if benchmark.language == 'jvm':
```
Comment (DonggeLiu) [inline run_one_experiment.py:253]: I just noticed this, should it be `elif benchmark.language == 'jvm':`?
Same for other conditions below.

**Review by DonggeLiu — APPROVED** [review]

File: data_prep/introspector.py (hunk 1 above)
Comment (DavidKorczynski) [reply data_prep/introspector.py]: Good point -- I adjusted so it uses it's own function, but is called by way of `populate_benchmarks_using_introspector` as this seemed a nice abstraction

File: run_all_experiments.py (hunk 4 above)
Comment (DavidKorczynski) [reply run_all_experiments.py]: Reverted the change!

File: run_one_experiment.py (hunk 5 above)
Comment (DavidKorczynski) [reply run_one_experiment.py]: I'm indifferent -- fixed it up to use `elif`s

File: experiment/benchmark.py (hunk 2 above)
Comment (DavidKorczynski) [reply experiment/benchmark.py:143]: Yep, will do

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-comparisonasdf12 -m vertex_ai_gemini-1-5 -b minor-for-ci

**Comment by DavidKorczynski** [comment]
/gcbrun exp -n dk-comparisonasfdf12 -m vertex_ai_gemini-1-5 -b minor-for-ci

**Pull request merged**